drivers/at86rf2xx/at86rf2xx_rf_ops.c

```diff
@@ -0,0 +1,944 @@
+            at86rf2xx_reg_write(dev, (AT86RF2XX_REG__IEEE_ADDR_0 + i),
+                    ext_addr->uint8[IEEE802154_LONG_ADDRESS_LEN - 1 - i]);
+                DEBUG("%02x", ext_addr->uint8[i]);
+        }
+        DEBUG("\n");
+        break;
+    case IEEE802154_AF_PANID: {
+        le_uint16_t le_pan = byteorder_btols(byteorder_htons(*pan_id));
+        at86rf2xx_reg_write(dev, AT86RF2XX_REG__PAN_ID_0, le_pan.u8[0]);
+        at86rf2xx_reg_write(dev, AT86RF2XX_REG__PAN_ID_1, le_pan.u8[1]);
+        DEBUG("PANID: %04x\n", *pan_id);
+    }
+        break;
+    case IEEE802154_AF_PAN_COORD:
+        mutex_unlock(&dev->lock);
+        return -ENOTSUP;
+    }
+
+    mutex_unlock(&dev->lock);
+    return 0;
+}
+
+static int _config_src_addr_match(ieee802154_dev_t *hal, ieee802154_src_match_t cmd, const void *value)
+{
+    at86rf2xx_t *dev = hal->priv;
+    int res;
+    mutex_lock(&dev->lock);
+    switch(cmd) {
+        case IEEE802154_SRC_MATCH_EN: {
+            const bool en = *((const bool*) value);
+            at86rf2xx_set_option(dev, AT86RF2XX_OPT_ACK_PENDING, en);
+            res = 0;
+            break;
+        }
+        default:
+            res = -ENOTSUP;
+    }
+    mutex_unlock(&dev->lock);
+    return res;
+}
+
+static int _set_frame_filter_mode(ieee802154_dev_t *hal, ieee802154_filter_mode_t mode)
+{
+    at86rf2xx_t *dev = hal->priv;
+    bool promisc = false;
+    switch(mode) {
+        case IEEE802154_FILTER_ACCEPT:
+            promisc = false;
+            break;
+        case IEEE802154_FILTER_PROMISC:
+            promisc = true;
+            break;
+        case IEEE802154_FILTER_ACK_ONLY:
+        case IEEE802154_FILTER_SNIFFER:
+            return -ENOTSUP;
+    }
+
+    DEBUG("[at86rf2xx] set promiscuous to %i\n", promisc);
+    mutex_lock(&dev->lock);
+    at86rf2xx_set_option(dev, AT86RF2XX_OPT_PROMISCUOUS, promisc);
+    mutex_unlock(&dev->lock);
+    return 0;
+}
+
+int at86rf2xx_init(at86rf2xx_t *dev, const at86rf2xx_params_t *params, ieee802154_dev_t *hal, ieee802154_sec_context_t *sec, void (*cb)(void*), void *ctx)
+{
+    uint8_t tmp;
+    (void) tmp;
+
+    (void) at86rf2xx_periph;
+    mutex_init(&dev->lock);
+    mutex_lock(&dev->lock);
+#if AT86RF2XX_IS_PERIPH
+    (void) params;
+    /* set all interrupts off */
+    at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, 0x00);
+    at86rf2xx_periph = hal;
+#else
+    /* initialize device descriptor */
+    dev->params = *params;
+#endif
+
+    if (!AT86RF2XX_IS_PERIPH) {
+        at86rf2xx_spi_init(dev, cb, ctx);
+    }
+
+    at86rf2xx_hardware_reset(dev);
+
+    /* test if the device is responding */
+    if (at86rf2xx_reg_read(dev, AT86RF2XX_REG__PART_NUM) != AT86RF2XX_PARTNUM) {
+        DEBUG("[at86rf2xx] error: unable to read correct part number\n");
+        mutex_unlock(&dev->lock);
+        return -ENOTSUP;
+    }
+
+    hal->driver = &at86rf2xx_ops;
+    hal->priv = dev;
+    (void) sec;
+#if IS_USED(MODULE_IEEE802154_SECURITY) && \
```

Comment:
If this is set here very early, it will be overridden in `netdev_submac_driver::_init()` -> `netdev_ieee802154_reset()` -> `ieee802154_sec_init()`. So now this function should check if it was set already.

Reply:
Addressed in [6b88935](https://github.com/RIOT-OS/RIOT/pull/19015/commits/6b8893514bd52c585212dea9ba9e57f24ac24396).

However, this implicitly assumes that the device descriptor was zeroed before calling `_init` (either via `memset` or by global allocation). This will work for all RIOT examples (since device descriptors are globally allocated), but some applications that use stack may not `memset` the device descriptor to zero.

To avoid any kind of confusion, I propose to set `cipher_ops` before calling the driver `_init` function and remove the assignment in `ieee802154_sec_init`. Although it adds a bit of boilerplate code, IMO it makes it clear that a device may use the default `cipher_ops`.

Reply:
any opinions on this?

Reply:
I am for this
<details>
<summary>patch</summary>

```diff
diff --git a/drivers/at86rf2xx/at86rf2xx_rf_ops.c b/drivers/at86rf2xx/at86rf2xx_rf_ops.c
index da2ad72814..370b4e34d3 100644
--- a/drivers/at86rf2xx/at86rf2xx_rf_ops.c
+++ b/drivers/at86rf2xx/at86rf2xx_rf_ops.c
@@ -105,6 +105,26 @@ static const ieee802154_radio_cipher_ops_t _at86rf2xx_cipher_ops = {
 #endif /* IS_USED(MODULE_AT86RF2XX_AES_SPI) && \
           IS_USED(MODULE_IEEE802154_SECURITY) */
 
+static const ieee802154_radio_cipher_ops_t *_get_cipher_ops(const ieee802154_dev_t *hal)
+{
+#if IS_USED(MODULE_AT86RF2XX_AES_SPI) && \
+    IS_USED(MODULE_IEEE802154_SECURITY)
+    return &_at86rf2xx_cipher_ops;
+#endif
+    (void)hal;
+    return NULL;
+}
+
+static void *_get_cipher_ops_ctx(const ieee802154_dev_t *hal)
+{
+#if IS_USED(MODULE_AT86RF2XX_AES_SPI) && \
+    IS_USED(MODULE_IEEE802154_SECURITY)
+    return hal->priv;
+#endif
+    (void)hal;
+    return NULL;
+}
+
 static int _write(ieee802154_dev_t *hal, const iolist_t *psdu)
 {
     DEBUG("at86rf2xx_rf_ops: write\n");
@@ -626,11 +646,6 @@ int at86rf2xx_init(at86rf2xx_t *dev, const at86rf2xx_params_t *params, ieee80215
     hal->driver = &at86rf2xx_ops;
     hal->priv = dev;
     (void) sec;
-#if IS_USED(MODULE_IEEE802154_SECURITY) && \
-    IS_USED(MODULE_AT86RF2XX_AES_SPI)
-    sec->dev.cipher_ops = &_at86rf2xx_cipher_ops;
-    sec->dev.ctx = dev;
-#endif
 
     if (!IS_ACTIVE(AT86RF2XX_BASIC_MODE)) {
         at86rf2xx_set_option(dev, AT86RF2XX_OPT_AUTOACK, true);
@@ -806,6 +821,8 @@ static const ieee802154_radio_ops_t at86rf2xx_ops = {
     .config_addr_filter = _config_addr_filter,
     .config_src_addr_match = _config_src_addr_match,
     .set_frame_filter_mode = _set_frame_filter_mode,
+    .get_cipher_ops = _get_cipher_ops,
+    .get_cipher_ops_ctx = _get_cipher_ops_ctx,
 };
 
 #if AT86RF2XX_IS_PERIPH
diff --git a/drivers/netdev_ieee802154_submac/netdev_ieee802154_submac.c b/drivers/netdev_ieee802154_submac/netdev_ieee802154_submac.c
index 1e6bf71d78..6438a66178 100644
--- a/drivers/netdev_ieee802154_submac/netdev_ieee802154_submac.c
+++ b/drivers/netdev_ieee802154_submac/netdev_ieee802154_submac.c
@@ -354,6 +354,14 @@ static int _init(netdev_t *netdev)
     /* This function already sets the PAN ID to the default one */
     netdev_ieee802154_reset(netdev_ieee802154);
 
+#if IS_USED(MODULE_IEEE802154_SECURITY)
+    const ieee802154_radio_cipher_ops_t *cipher_ops;
+    if ((cipher_ops = ieee802154_radio_get_cipher_ops(&submac->dev))) {
+        netdev_ieee802154->sec_ctx.dev.cipher_ops = cipher_ops;
+        netdev_ieee802154->sec_ctx.dev.ctx = ieee802154_radio_get_cipher_ops_ctx(&submac->dev);
+    }
+#endif
+
     uint16_t chan = submac->channel_num;
     int16_t tx_power = submac->tx_pow;
     static const netopt_enable_t ack_req =
diff --git a/sys/include/net/ieee802154/radio.h b/sys/include/net/ieee802154/radio.h
index c8f71afa90..5bbe6b3e60 100644
--- a/sys/include/net/ieee802154/radio.h
+++ b/sys/include/net/ieee802154/radio.h
@@ -40,6 +40,11 @@ extern "C" {
  */
 typedef struct ieee802154_radio_ops ieee802154_radio_ops_t;
 
+/**
+ * @brief Forward declaration of radio cipher ops structure.
+ */
+struct ieee802154_radio_cipher_ops;
+
 /**
  * @brief IEEE802.15.4 Radio capabilities
  *
@@ -775,6 +780,22 @@ struct ieee802154_radio_ops {
      */
     int (*config_src_addr_match)(ieee802154_dev_t *dev, ieee802154_src_match_t cmd,
                                  const void *value);
+
+    /**
+     * @brief Retrieve radio cipher ops
+     *
+     * @return Radio cipher ops
+     * @return NULL if device has no cipher ops
+     */
+    const struct ieee802154_radio_cipher_ops *(*get_cipher_ops)(const ieee802154_dev_t *dev);
+
+    /**
+     * @brief Retrieve radio cipher ops context
+     *
+     * @return Radio cipher ops context
+     * @return NULL if device has no cipher ops context
+     */
+    void *(*get_cipher_ops_ctx)(const ieee802154_dev_t *dev);
 };
 
 /**
@@ -1272,6 +1293,28 @@ static inline int ieee802154_radio_cca(ieee802154_dev_t *dev)
     return res;
 }
 
+/**
+ * @brief Retrieve radio cipher ops
+ *
+ * @return Radio cipher ops
+ * @return NULL if device has no cipher ops
+ */
+static inline const struct ieee802154_radio_cipher_ops *ieee802154_radio_get_cipher_ops(const ieee802154_dev_t *dev)
+{
+    return dev->driver->get_cipher_ops(dev);
+}
+
+/**
+ * @brief Retrieve radio cipher ops context
+ *
+ * @return Radio cipher ops context
+ * @return NULL if device has no cipher ops context
+ */
+static inline void *ieee802154_radio_get_cipher_ops_ctx(const ieee802154_dev_t *dev)
+{
+    return dev->driver->get_cipher_ops_ctx(dev);
+}
+
 /**
  * @brief Check if the device supports ACK timeout
  *
diff --git a/sys/net/gnrc/netif/init_devs/auto_init_at86rf2xx.c b/sys/net/gnrc/netif/init_devs/auto_init_at86rf2xx.c
index db0aa56673..4526acb7fd 100644
--- a/sys/net/gnrc/netif/init_devs/auto_init_at86rf2xx.c
+++ b/sys/net/gnrc/netif/init_devs/auto_init_at86rf2xx.c
@@ -56,12 +56,7 @@ void auto_init_at86rf2xx(void)
     for (unsigned i = 0; i < AT86RF2XX_NUM; i++) {
         LOG_DEBUG("[auto_init_netif] initializing at86rf2xx #%u\n", i);
 
-        ieee802154_sec_context_t *sec = NULL;
-#if IS_USED(MODULE_IEEE802154_SECURITY)
-        sec = &at86rf2xx_netdev[i].dev.sec_ctx;
-#endif
-
-        at86rf2xx_init_event(&at86rf2xx_bhp[i], &at86rf2xx_params[i], &at86rf2xx_netdev[i].submac.dev, sec, EVENT_PRIO_HIGHEST);
+        at86rf2xx_init_event(&at86rf2xx_bhp[i], &at86rf2xx_params[i], &at86rf2xx_netdev[i].submac.dev, NULL, EVENT_PRIO_HIGHEST);
         netdev_register(&at86rf2xx_netdev[i].dev.netdev, NETDEV_AT86RF2XX, i);
         netdev_ieee802154_submac_init(&at86rf2xx_netdev[i]);
 #if defined(MODULE_GNRC_GOMACH)
diff --git a/sys/net/link_layer/ieee802154/security.c b/sys/net/link_layer/ieee802154/security.c
index 83782ba9e8..3271b86557 100644
--- a/sys/net/link_layer/ieee802154/security.c
+++ b/sys/net/link_layer/ieee802154/security.c
@@ -391,12 +391,10 @@ static void _ctr_mic(ieee802154_sec_context_t *ctx,
 
 void ieee802154_sec_init(ieee802154_sec_context_t *ctx)
 {
-    /* If context is not set by the device, set it to default */
-    if (!ctx->dev.cipher_ops) {
-        ctx->dev.cipher_ops = &ieee802154_radio_cipher_ops;
-        ctx->dev.ctx = ctx;
-    }
-
+    /* device driver can override this */
+    ctx->dev.cipher_ops = &ieee802154_radio_cipher_ops;
+    /* device driver can override this */
+    ctx->dev.ctx = ctx;
     /* MIC64 is the only mandatory security mode */
     ctx->security_level = IEEE802154_SEC_SCF_SECLEVEL_ENC_MIC64;
     ctx->key_id_mode = IEEE802154_SEC_SCF_KEYMODE_IMPLICIT;

```
</details>

Reply:
sure, makes sense

Reply:
Ok to not implement a dummy return `NULL` function for all transceivers not having crypto hardware support, the function pointers could be `NULL` instead.